apiserver/facades/client/action/operation.go

```diff
@@ -82,30 +82,45 @@ func (a *ActionAPI) enqueue(arg params.Actions) (string, params.ActionResults, e
 
 	tagToActionReceiver := common.TagToActionReceiverFn(a.state.FindEntity)
 	response := params.ActionResults{Results: make([]params.ActionResult, len(arg.Actions))}
+	errorRecorded := false
 	for i, action := range arg.Actions {
 		currentResult := &response.Results[i]
 		actionReceiver := action.Receiver
+		var (
+			actionErr    error
+			enqueued     state.Action
+			receiver     state.ActionReceiver
+			receiverName string
+		)
 		if strings.HasSuffix(actionReceiver, "leader") {
 			app := strings.Split(actionReceiver, "/")[0]
-			receiverName, err := getLeader(app)
-			if err != nil {
-				currentResult.Error = common.ServerError(err)
-				continue
+			receiverName, actionErr = getLeader(app)
+			if actionErr != nil {
+				currentResult.Error = common.ServerError(actionErr)
+				goto failOperation
 			}
 			actionReceiver = names.NewUnitTag(receiverName).String()
 		}
-		receiver, err := tagToActionReceiver(actionReceiver)
-		if err != nil {
-			currentResult.Error = common.ServerError(err)
-			continue
+		receiver, actionErr = tagToActionReceiver(actionReceiver)
+		if actionErr != nil {
+			currentResult.Error = common.ServerError(actionErr)
+			goto failOperation
 		}
-		enqueued, err := receiver.AddAction(operationID, action.Name, action.Parameters)
-		if err != nil {
-			currentResult.Error = common.ServerError(err)
-			continue
+		enqueued, actionErr = receiver.AddAction(operationID, action.Name, action.Parameters)
+		if actionErr != nil {
+			currentResult.Error = common.ServerError(actionErr)
+			goto failOperation
 		}
-
 		response.Results[i] = common.MakeActionResult(receiver.Tag(), enqueued, false)
+		continue
+
+	failOperation:
+		// If we failed to enqueue the action, record the error on the operation.
+		if !errorRecorded {
+			errorRecorded = true
+			err = a.model.FailOperation(operationID, actionErr)
```

Comment:
so do we only record the first action error to the operation?

Reply:
yeah, 99% of the time it will be just one action run across several targets. that's what the juju run cli invokes